### PR TITLE
Introduce support for the BulkImport serialization to JSON, Flatbuffe…

### PR DESCRIPTION
**Pull request by ftomassetti**
This is needed to support the bulkImport operation, proposed in https://github.com/LionWeb-io/lionweb-repository/pull/81

### CODE REVIEW

**Comment by ftomassetti** [comment]
> Any chance to keep this separated from the official bulk protocol?

Given these structures are useful for communication with the repo, I wanted to put them in LionWeb Kotlin. 

HowevertThe problem is that we need to introduce the BulkImport data structure for both protobuf and flatbuffers, and to do that we need to extend the schemas we already have for serializing the schema previously defined for the chunk. This is why this ended up being in this project.

Would it be enough to create a different serialization class just for the bulk import, but keeping it within this project?

**Comment by enikao** [comment]
That's a start.

Do you intend to send the messages through the same channel? If not, it might work to move the new messages into a different file and import the other one, creating an independent set of source files (which might even be in another subproject).

**Comment by ftomassetti** [comment]
> Do you intend to send the messages through the same channel? 

Yes, it would be used in HTTP calls to the LionWeb Repository, as for the normal inserts or retrieval

**Comment by enikao** [comment]
Ok, but that would be a different http request, i.e. a new connection -- then using different messages should work? Would be different if we connected e.g. via websocket, then all messages would go through the same connection.

**Comment by ftomassetti** [comment]
> Ok, but that would be a different http request, i.e. a new connection -- then using different messages should work? Would be different if we connected e.g. via websocket, then all messages would go through the same connection.

If by different messages we mean different classes, then it would mean writing some code twice, so I would avoid that if possible. The BulkImport data structure is a tiny wrapper that reuse the node definition and all of its parts (metapointers, features, etc.)

I started by creating separate Serialization classes to handle the BulkImport data structure.

**Comment by enikao** [comment]
I was thinking like:

* LionWeb.core
  * lionweb.protobuf
  * lionweb.flatbuf
  * (generated classes)

* LionWeb.extension
  * lionweb-extension.protobuf
    import ../LionWeb.core/lionweb.protobuf
    ...  
  * lionweb-extension.flatbuf
    import ../LionWeb.core/lionweb.flatbuf
    ...  
  * (generated classes)


But if this is hard to do a different serializer is a good middle ground.

**Comment by ftomassetti** [comment]
Ok, I think I may have found a way to use imports for both protobufs and flatbuffers and do not regenerate the imported classes. So I created the extensions module as suggested

**Comment by ftomassetti** [comment]
Now that https://github.com/LionWeb-io/lionweb-repository/pull/81 has been merged, this one can be reviewed

File: core/src/main/java/io/lionweb/lioncore/java/serialization/ProtoBufSerialization.java [hunk 1]
```diff
@@ -160,11 +160,13 @@ public byte[] serializeToByteArray(SerializedChunk serializedChunk) {
     return serialize(serializedChunk).toByteArray();
   }
 
-  private class SerializeHelper {
-    final Map<MetaPointer, Integer> metaPointers = new HashMap<>();
-    final Map<String, Integer> strings = new HashMap<>();
+  protected class SerializeHelper {
+    public final Map<MetaPointer, Integer> metaPointers = new HashMap<>();
```
Comment (enikao) [inline core/src/main/java/io/lionweb/lioncore/java/serialization/ProtoBufSerialization.java:164]: `public` sounds suspicious for fields (same for next line)

File: extensions/build.gradle.kts [hunk 2]
```diff
@@ -0,0 +1,119 @@
+import com.vanniktech.maven.publish.SonatypeHost
+import java.net.URI
+
+plugins {
+    id("java-library")
+    id("signing")
+    id("com.github.johnrengelman.shadow") version "7.1.1"
```
Comment (enikao) [inline extensions/build.gradle.kts:7]: didn't we keep all versions in `settings.gradle` or `gradle.properties`?

File: extensions/build.gradle.kts [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+    extendsFrom(configurations.testImplementation.get())
+}
+
+dependencies {
+    implementation(project(":core"))
+
+    // Use JUnit test framework.
+    testImplementation(libs.junit)
+
+    implementation("com.google.protobuf:protobuf-java:4.27.2")
```
Comment (enikao) [inline extensions/build.gradle.kts:30]: another set of version numbers

File: extensions/build.gradle.kts [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+
+tasks.jacocoTestReport {
+    dependsOn(tasks.test) // tests are required to run before generating the report
+}
+
+protobuf {
+    protoc {
+        protoc {
+            // Apple Silicon processor would look for an unexisting platform-specific variant
+            artifact = "com.google.protobuf:protoc:4.27.2" + if (osdetector.os == "osx") ":osx-x86_64" else ""
```
Comment (enikao) [inline extensions/build.gradle.kts:107]: another version

File: extensions/build.gradle.kts [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+    sourceCompatibility = JavaVersion.toVersion(jvmVersion)
+    targetCompatibility = JavaVersion.toVersion(jvmVersion)
+}
+
+tasks.jacocoTestReport {
+    dependsOn(tasks.test) // tests are required to run before generating the report
+}
+
+protobuf {
+    protoc {
```
Comment (enikao) [inline extensions/build.gradle.kts:104]: really two nested `protoc` blocks?

File: extensions/src/main/java/io/lionweb/serialization/flatbuffers/FBAttachPoint.java [hunk 6]
```diff
@@ -0,0 +1,116 @@
+// automatically generated by the FlatBuffers compiler, do not modify
```
Comment (enikao) [inline extensions/src/main/java/io/lionweb/serialization/flatbuffers/FBAttachPoint.java:1]: maybe move generated classes into a `.gen.` subpackage?

File: core/src/main/java/io/lionweb/lioncore/java/serialization/ProtoBufSerialization.java (hunk 1 above)
Comment (ftomassetti) [reply core/src/main/java/io/lionweb/lioncore/java/serialization/ProtoBufSerialization.java:164]: Changed

File: extensions/build.gradle.kts (hunk 5 above)
Comment (ftomassetti) [reply extensions/build.gradle.kts:106]: Removed

File: extensions/build.gradle.kts (hunk 4 above)
Comment (ftomassetti) [reply extensions/build.gradle.kts]: I added a versions catalog and put it there

File: extensions/build.gradle.kts (hunk 3 above)
Comment (ftomassetti) [reply extensions/build.gradle.kts]: I added a versions catalog and put it there

File: extensions/build.gradle.kts (hunk 2 above)
Comment (ftomassetti) [reply extensions/build.gradle.kts]: I added a versions catalog and put it there

File: extensions/src/main/java/io/lionweb/serialization/extensions/ExtraFlatBuffersSerialization.java [hunk 7]
```diff
@@ -0,0 +1,184 @@
...
+import io.lionweb.serialization.flatbuffers.gen.*;
+import java.util.HashMap;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Map;
+
+/** It contains the logic to serialize non-standard messages. */
+public class ExtraFlatBuffersSerialization extends FlatBuffersSerialization {
+
+  public byte[] serializeBulkImport(BulkImport bulkImport) {
```
Comment (enikao) [inline extensions/src/main/java/io/lionweb/serialization/extensions/ExtraFlatBuffersSerialization.java:23]: This is a pretty long method, consider splitting it up.

File: settings.gradle [hunk 8]
```diff
@@ -1,8 +1,13 @@
+pluginManagement {
+    plugins {
+        id 'org.jetbrains.kotlin.jvm' version '2.0.0'
```
Comment (enikao) [inline settings.gradle:3]: Why do we use kotlin in lw-java?

Comment (ftomassetti) [reply settings.gradle]: At some point I must have added a Kotlin file by mistake and IDEA must have added this automatically. I removed it now

File: extensions/src/main/java/io/lionweb/serialization/extensions/ExtraFlatBuffersSerialization.java (hunk 7 above)
Comment (ftomassetti) [reply extensions/src/main/java/io/lionweb/serialization/extensions/ExtraFlatBuffersSerialization.java:23]: I extracted two methods from this one

**Comment by ftomassetti** [comment]
Thank you @enikao !

**Pull request merged**